src/themespresetmanager.cpp

```diff
@@ -0,0 +1,288 @@
+{
+}
+
+bool RepoLocalFileSystemSource::loadThemesDb()
+{
+
+  if (!isDir(_pathToFolder)){
+    return false;
+  }
+  for(const auto & entry : std::filesystem::directory_iterator(_pathToFolder)){
+    std::string configPath = checkIsDirHaveConfig(entry.path());
+    if  (configPath != ""){
+      _themesDb[parseThemeAliasFromFilename(entry.path())] = configPath;
+    }
+  }
+  if (_themesDb.empty()){
+    return false;
+  }
+  return true;
+
+}
+
+
+std::string RepoLocalFileSystemSource::checkIsDirHaveConfig(std::string path)
+{
+  if (!isDir(path)){
+    return "";
+  }
+  for (const auto & entry : std::filesystem::directory_iterator(path)){
+
+    std::string filename = entry.path();
+    filename = filename.substr(filename.rfind("/") +1, filename.size()); // remove dirs and get simply filename
+    if (filename.find(THEME_CONFIG_END)!=filename.npos){
+      return entry.path();
+    }
+  }
+  return ""; // cant find config
+
+}
+
+SystemGitRepoSource::SystemGitRepoSource(std::string  pathToNewGitRepo, std::string cloneUrlToGitRepo):RepoLocalFileSystemSource(pathToNewGitRepo),_repoUrl(cloneUrlToGitRepo)
+{
+  git_libgit2_init();
+}
+
+SystemGitRepoSource::SystemGitRepoSource(std::string pathToGitAlreadyClonedRepo):RepoLocalFileSystemSource(pathToGitAlreadyClonedRepo)
+{
+  git_libgit2_init();
+}
+
+SystemGitRepoSource::~SystemGitRepoSource()
+{
+  git_libgit2_shutdown();
+}
+
+bool SystemGitRepoSource::loadThemesDb()
+{
+  if (!isDir(_pathToFolder)){
+    if (!mkdir(_pathToFolder)){
+      std::cout << "SystemGitRepoSource::path dir repo not exists and cant create" << std::endl;
+      return false;
+    }
+  }
+  if (isGettedPathGitRepo(_pathToFolder)){
+    if (openRepo(_pathToFolder)){
+      if (pullChanges()){
+        return RepoLocalFileSystemSource::loadThemesDb();
+      }
+      return false;
+    }
+    return false;
+  }
+  else{
+    if (_repoUrl != ""){
+      std::cout << "cloning" << std::endl;
+      if (git_clone(&_repo, _repoUrl.c_str(), _pathToFolder.c_str(), NULL)!=0){
+        std::cout << "SystemGitRepoSource::error while git clone(init url): "<<git_error_last()->message << std::endl;
+        return false;
+      }
+      return RepoLocalFileSystemSource::loadThemesDb();
+    }
+  }
+  return false;
+
+
+}
+
+bool SystemGitRepoSource::pullChanges() // maybe need to change this
+{
+  git_remote *remote;
+  if(git_remote_lookup(&remote, _repo, "origin")!=0){
+    std::cout << "SystemGitRepoSource::error while git remote lookup (origin): "<<git_error_last()->message << std::endl;
+    return false;
+  }
+  if (git_remote_fetch(remote,NULL,NULL,"fetch")!=0){
+    std::cout << "SystemGitRepoSource::error while git remote fetch (origin): "<<git_error_last()->message << std::endl;
+    return false;
+  }
+  git_reference * mainBranch = nullptr;
+  if (git_reference_lookup(&mainBranch, _repo, "refs/remotes/origin/main")!=0){
```

Comment:
Hardcoding "main" ref for repositories feels bad. It should be specified in the same place the repository URI is.